File: frontend/src/services/promptService.ts
import axiosInstance from '../lib/axiosInstance';

const LLM_TIMEOUT_MS = 5 * 60 * 1000;

export interface AnalyzeResponse {
  category: string;
  qualityScore: number;
  estimatedTokens: number;
  costEstimates: Record<string, number>;
  recommendations: {
    bestOverall: string;
    bestValue: string;
    cheapest: string;
  };
}

export interface OptimizeResponse {
  optimizedPrompt: string;
  tokensBefore: number;
  tokensAfter: number;
  savingsPercent: number;
  costBefore: number;
  costAfter: number;
}

export interface PromptDto {
  id: number;
  originalPrompt: string;
  optimizedPrompt?: string;
  category: string;
  qualityScore: number;
  tokensBefore?: number;
  tokensAfter?: number;
  costBefore?: number;
  costAfter?: number;
  recommendedModel?: string;
  optimizationMode?: string;
  createdAt: string;
}

export interface StatisticsDto {
  totalPrompts: number;
  totalTokensSaved: number;
  totalMoneySaved: number;
  averageQualityScore: number;
}

export const promptService = {
  analyze: (prompt: string): Promise<{ data: AnalyzeResponse }> =>
    axiosInstance.post('/prompts/analyze', { prompt }, { timeout: LLM_TIMEOUT_MS }),

  optimize: (originalPrompt: string, mode: string): Promise<{ data: OptimizeResponse }> =>
    axiosInstance.post('/prompts/optimize', { originalPrompt, mode }, { timeout: LLM_TIMEOUT_MS }),

  save: (data: object) =>
    axiosInstance.post('/prompts', data),

  getHistory: (page = 0, size = 10) =>
    axiosInstance.get(`/prompts?page=${page}&size=${size}`),

  delete: (id: number) =>
    axiosInstance.delete(`/prompts/${id}`),

  getStats: (): Promise<{ data: StatisticsDto }> =>
    axiosInstance.get('/prompts/stats'),
};

File: backend/advisor/src/main/java/com/smartllm/advisor/service/PromptService.java
package com.smartllm.advisor.service;

import com.smartllm.advisor.dto.PromptDto;
import com.smartllm.advisor.dto.PromptSaveRequest;
import com.smartllm.advisor.entity.Prompt;
import com.smartllm.advisor.entity.User;
import com.smartllm.advisor.repository.PromptRepository;
import com.smartllm.advisor.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;

@Service
public class PromptService {

    @Autowired private PromptRepository promptRepository;
    @Autowired private UserRepository userRepository;

    public PromptDto save(PromptSaveRequest request) {
        User user = getCurrentUser();

        Prompt prompt = Prompt.builder()
                .originalPrompt(request.getOriginalPrompt())
                .optimizedPrompt(request.getOptimizedPrompt())
                .category(request.getCategory())
                .qualityScore(request.getQualityScore())
                .tokensBefore(request.getTokensBefore())
                .tokensAfter(request.getTokensAfter())
                .costBefore(request.getCostBefore() != null
                        ? BigDecimal.valueOf(request.getCostBefore()) : null)
                .costAfter(request.getCostAfter() != null
                        ? BigDecimal.valueOf(request.getCostAfter()) : null)
                .recommendedModel(request.getRecommendedModel())
                .optimizationMode(request.getOptimizationMode())
                .user(user)
                .build();

        Prompt saved = promptRepository.save(prompt);
        return toDto(saved);
    }

    public Page<PromptDto> getHistory(int page, int size) {
        User user = getCurrentUser();
        Pageable pageable = PageRequest.of(page, size);
        return promptRepository.findByUserIdOrderByCreatedAtDesc(user.getId(), pageable)
                .map(this::toDto);
    }

    public void delete(Long id) {
        User user = getCurrentUser();
        Prompt prompt = promptRepository.findByIdAndUserId(id, user.getId())
                .orElseThrow(() -> new RuntimeException("Prompt not found or access denied"));
        promptRepository.delete(prompt);
    }

    private User getCurrentUser() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));
    }

    private PromptDto toDto(Prompt p) {
        return PromptDto.builder()
                .id(p.getId())
                .originalPrompt(p.getOriginalPrompt())
                .optimizedPrompt(p.getOptimizedPrompt())
                .category(p.getCategory())
                .qualityScore(p.getQualityScore())
                .tokensBefore(p.getTokensBefore())
                .tokensAfter(p.getTokensAfter())
                .costBefore(p.getCostBefore() != null ? p.getCostBefore().doubleValue() : null)
                .costAfter(p.getCostAfter() != null ? p.getCostAfter().doubleValue() : null)
                .recommendedModel(p.getRecommendedModel())
                .optimizationMode(p.getOptimizationMode())
                .createdAt(p.getCreatedAt())
                .build();
    }
}